Merge upstream 5.5.57 into Percona Server

percona-ysorokin opened the pull request.

percona-ysorokin commented:
Jenkins build link (all blue)
http://jenkins.percona.com/view/PS%205.5/job/percona-server-5.5-param/1588/

laurynas-biveinis commented:
Name the testcase main.bug84640, use actual debug sync points from the code: mi_open_datafile_before_open became before_opening_datafile. If you feel like it, feel free to extend to INDEX DIRECTORY too, but this is not critical

```
diff --git a/mysql-test/r/myisam_symlink.result b/mysql-test/r/myisam_symlink.result
new file mode 100644
index 00000000000..375781a5298
--- /dev/null
+++ b/mysql-test/r/myisam_symlink.result
@@ -0,0 +1,27 @@
+#
+# Tests for MyISAM handling of symlinks
+#
+#
+# Bug 1654256: symlinks in CREATE TABLE ... DATA DIRECTORY, last path component
+#
+CREATE TABLE t1 (a INT) ENGINE=MyISAM DATA DIRECTORY='custom_data_directory';
+INSERT INTO t1 VALUES (1), (2);
+FLUSH TABLES;
+SET DEBUG_SYNC="mi_open_datafile_before_open SIGNAL open_ready WAIT_FOR finish_open";
+SELECT * FROM t1;
+SET DEBUG_SYNC="now WAIT_FOR open_ready";
+SET DEBUG_SYNC="now SIGNAL finish_open";
+ERROR HY000: File 'custom_data_directory/t1.MYD' not found (Errcode: 62)
+DROP TABLE t1;
+#
+# Bug 1654256: symlinks in CREATE TABLE ... DATA DIRECTORY, middle path component
+#
+CREATE TABLE user(a INT) ENGINE=MyISAM DATA DIRECTORY='custom_data_directory';
+INSERT INTO user VALUES (1), (2);
+FLUSH TABLES;
+SET DEBUG_SYNC="mi_open_datafile_before_open SIGNAL open_ready WAIT_FOR finish_open";
+SELECT * FROM user;
+SET DEBUG_SYNC="now WAIT_FOR open_ready";
+SET DEBUG_SYNC="now SIGNAL finish_open";
+ERROR HY000: Got error 140 from storage engine
+DROP TABLE user;
diff --git a/mysql-test/t/myisam_symlink.test b/mysql-test/t/myisam_symlink.test
new file mode 100644
index 00000000000..aa704b2b8c0
--- /dev/null
+++ b/mysql-test/t/myisam_symlink.test
@@ -0,0 +1,88 @@
+--source include/have_debug_sync.inc
+--source include/have_symlink.inc
+--source include/count_sessions.inc
+
+--echo #
+--echo # Tests for MyISAM handling of symlinks
+--echo #
+
+--let $datadir= `SELECT @@datadir`
+--let $tmpdir= `SELECT @@tmpdir`
+--let $custom_data_directory= $tmpdir/foo
+
+--echo #
+--echo # Bug 1654256: symlinks in CREATE TABLE ... DATA DIRECTORY, last path component
+--echo #
+
+--mkdir $custom_data_directory
+
+--replace_result $custom_data_directory custom_data_directory
+eval CREATE TABLE t1 (a INT) ENGINE=MyISAM DATA DIRECTORY='$custom_data_directory';
+
+INSERT INTO t1 VALUES (1), (2);
+
+FLUSH TABLES;
+
+--connect(con1,localhost,root,,)
+
+SET DEBUG_SYNC="mi_open_datafile_before_open SIGNAL open_ready WAIT_FOR finish_open";
+
+send SELECT * FROM t1;
+
+--connection default
+
+SET DEBUG_SYNC="now WAIT_FOR open_ready";
+
+--remove_file $custom_data_directory/t1.MYD
+--exec ln -s $datadir/mysql/user.MYD $custom_data_directory/t1.MYD
+
+SET DEBUG_SYNC="now SIGNAL finish_open";
+
+--connection con1
+
+--replace_result $custom_data_directory custom_data_directory
+--error 29
+reap;
+
+--disconnect con1
+--connection default
+
+DROP TABLE t1;
+
+--echo #
+--echo # Bug 1654256: symlinks in CREATE TABLE ... DATA DIRECTORY, middle path component
+--echo #
+
+--replace_result $custom_data_directory custom_data_directory
+eval CREATE TABLE user(a INT) ENGINE=MyISAM DATA DIRECTORY='$custom_data_directory';
+
+INSERT INTO user VALUES (1), (2);
+
+FLUSH TABLES;
+
+--connect(con1,localhost,root,,)
+
+SET DEBUG_SYNC="mi_open_datafile_before_open SIGNAL open_ready WAIT_FOR finish_open";
+send SELECT * FROM user;
+
+--connection default
+
+SET DEBUG_SYNC="now WAIT_FOR open_ready";
+
+--exec rm -rf $custom_data_directory
+--exec ln -s "$datadir"mysql $custom_data_directory
+
+SET DEBUG_SYNC="now SIGNAL finish_open";
+
+--connection con1
+
+--error ER_GET_ERRNO
+reap;
+--disconnect con1
+--connection default
+
+DROP TABLE user;
+
+--remove_file $custom_data_directory
+
+--source include/wait_until_count_sessions.inc
```

percona-ysorokin commented:
Additional Jenkins build link with just one `main.bug84640` MTR test case
http://jenkins.percona.com/view/PS%205.5/job/percona-server-5.5-param/1589/

percona-ysorokin commented:
MTR test case `main.myisam_symlink` renamed to `main.bug84640.`.
`$custom_data_directory` initialized with more unique `$tmpdir/bug84640`
Added a few `SET DEBUG_SYNC= 'RESET';` statements.
Test case extended with `CREATE TABLE ... INDEX DIRECTORY` section.
Added more leftovers cleanup between sections.

laurynas-biveinis commented on mysql-test/t/bug84640.test:
```diff
@@ -0,0 +1,142 @@
+--connection con1
+
+--replace_result $custom_data_directory custom_data_directory
+--error 29
```
ER_ code?

percona-ysorokin replied:
Yes, I noticed this as well, but unfortunately there is no symbolic names for codes with values less than 1000.
